### PR TITLE
ccgx: add hybrid dock in quirk, add hpi cmd retry

### PR DESCRIPTION
**Pull request by chlee75**
Type of pull request:
- [ ] Code fix
  1. Fix and Add Lenovo Hybrid Dock meta data in quirk file
  2. Add HPI command retry to prevent command fail due to other events in Hybrid Dock

### CODE REVIEW

File: plugins/ccgx/fu-ccgx-hpi-device.c [hunk 1]
```diff
@@ -70,12 +73,24 @@ fu_ccgx_hpi_device_to_string (FuDevice *device, guint idt, GString *str)
...
 
 typedef struct {
-	guint8			 mode;
-	guint16			 addr;
-	guint8			*buf;
-	gsize			 bufsz;
+	guint8		mode;
+	guint16		addr;
+	guint8		*buf;
+	gsize		bufsz;
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-hpi-device.c:79]: indenting is wrong; fields without the `*` need a single space after the tab.

File: plugins/ccgx/fu-ccgx-hpi-device.c [hunk 2]
```diff
@@ -696,16 +711,21 @@ fu_ccgx_hpi_device_clear_all_events (FuCcgxHpiDevice *self,
 }
 
 static gboolean
-fu_ccgx_hpi_validate_fw (FuCcgxHpiDevice *self, guint8 fw_index, GError **error)
+fu_ccgx_hpi_validate_fw_cb (FuDevice *device, gpointer user_data, GError **error)
 {
+	FuCcgxHpiDevice *self = FU_CCGX_HPI_DEVICE (device);
+	guint8* fw_index = (guint8*) user_data;
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-hpi-device.c:717]: space star plaese.

File: plugins/ccgx/fu-ccgx-hpi-device.c [hunk 3]
```diff
@@ -909,13 +973,31 @@ fu_ccgx_hpi_read_flash (FuCcgxHpiDevice *self,
...
+			GError **error)
+{
+	FuCcgxHpiFlashReadRetryHelper helper = {
+		.addr = addr,
+		.buf = buf,
+		.bufsz = bufsz,
+	};
+	return fu_device_retry (FU_DEVICE (self),
+			        fu_ccgx_hpi_read_flash_cb,
+			        HPI_CMD_FLASH_READ_WRITE_RETRY_CNT,
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-hpi-device.c:997]: `_WRITE_RETRY_CNT`?

File: plugins/ccgx/fu-ccgx-hpi-device.c [hunk 4]
```diff
@@ -1467,17 +1549,21 @@ fu_ccgx_hpi_device_close (FuUsbDevice *device, GError **error)
...
-		g_set_error (error,
-			     FWUPD_ERROR,
-			     FWUPD_ERROR_NOT_SUPPORTED,
-			     "cannot release interface: %s",
-			     error_local->message);
-		return FALSE;
-	}
+	
+	/* do not close handle when device restarts */
+	if (fu_device_get_status (FU_DEVICE (device)) != FWUPD_STATUS_DEVICE_RESTART) {
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-hpi-device.c:1554]: maybe just

    if (fu_device_get_status (FU_DEVICE (device)) == FWUPD_STATUS_DEVICE_RESTART) {
        return TRUE;

?

**Comment by chlee75** [comment]
@hughsie 
I fixed code rule.

**Comment by chlee75** [comment]
@hughsie 
> no, it's for the items that don'thave * pointers. It's to make the addr, buf, bufsz line up vertically.

 Please check vertical lineup

File: plugins/ccgx/fu-ccgx-hpi-device.c [hunk 5]
```diff
@@ -70,12 +74,24 @@ fu_ccgx_hpi_device_to_string (FuDevice *device, guint idt, GString *str)
...
 
 typedef struct {
-	guint8			 mode;
-	guint16			 addr;
-	guint8			*buf;
-	gsize			 bufsz;
+	guint8 		 mode;
+	guint16 	 addr;
+	guint8 		*buf;
+	gsize 		 bufsz;
```
Comment (hughsie) [inline plugins/ccgx/fu-ccgx-hpi-device.c:80]: same here

Comment (chlee75) [reply plugins/ccgx/fu-ccgx-hpi-device.c:80]: @hughsie 
`same here`

In line 77 ~ 80, I think that mode,addr,buf,bufsz are vertically lineuped.
Am I wrong?

Comment (hughsie) [reply plugins/ccgx/fu-ccgx-hpi-device.c:80]: it's currently space tab space.

Comment (chlee75) [reply plugins/ccgx/fu-ccgx-hpi-device.c:80]: @hughsie 
`it's currently space tab space.`

Which one is right between tab + space or space + tab?
To lineup *, I think one space should be needed.

Comment (hughsie) [reply plugins/ccgx/fu-ccgx-hpi-device.c:80]: as a rule, space never comes before tab. It should be tab space for no pointers, and just tab for pointers.

**Comment by chlee75** [comment]
@hughsie 
`as a rule, space never comes before tab. It should be tab space for no pointers, and just tab for pointers.`

Which editor do you use?
I edited code in LVFS web and copy it to gedit. and I commit.
I am not sure whether this would be effective for LVFS coding rule or not?

**Comment by joakim-tjernlund** [comment]
> @hughsie
> `as a rule, space never comes before tab. It should be tab space for no pointers, and just tab for pointers.`
> 
> Which editor do you use?
> I edited code in LVFS web and copy it to gedit. and I commit.

gedit is not a code editor, you need an editor that understand C indent rules. Emacs is one but hard to learn. 
> I am not sure whether this would be effective for LVFS coding rule or not?

**Comment by hughsie** [comment]
> Which editor do you use?

Either vim or geany. The latter is easy to use and is like gedit. You can turn on showing whitespace in the preferences.

**Comment by chlee75** [comment]
@hughsie 

`Either vim or geany. The latter is easy to use and is like gedit. You can turn on showing whitespace in the preferences.`

I used geany but alignment display is different between geany and LVFS web.
What to fix it?

What about last commit? Is it right?

**Comment by superm1** [comment]
> > Which editor do you use?
> 
> Either vim or geany. The latter is easy to use and is like gedit. You can turn on showing whitespace in the preferences.

Just to throw another hat into the ring, Visual Studio code is also quite straightforward and can render whitespace for you.  It lets you set up other useful rules that match project preferences like this:
* Chop off trailing whitespace
* configure file to use spaces or tabs
* Match line endings (such as if you edit file on Windows instead)

Also if you install the C++ extension it will do structure detection (Intellisense - very similar to the Visual Studio intellisense you get in the full IDE) and code completion for you.

**Comment by hughsie** [comment]
> I used geany but alignment display is different between geany and LVFS web.

Don't worry how github formats it. I care about the editor :)

**Comment by chlee75** [comment]
@hughsie  @superm1 

I will try to setup best editor. Thanks a lot.

@hughsie 

`as a rule, space never comes before tab. It should be tab space for no pointers, and just tab for pointers`
Does my last commit (ccgx: make an alignment tab space for non ptr and tab for ptr) matches this rule?

**Pull request merged**